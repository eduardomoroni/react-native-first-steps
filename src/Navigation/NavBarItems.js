/* @flow */

import React from 'react'
import {
  StyleSheet,
  TouchableOpacity
} from 'react-native'
import { Colors, Metrics } from '../Styles/Themes/'
import Icon from 'react-native-vector-icons/FontAwesome'
import { navButtonColor } from '../Styles/NavBarStyle'

type ButtonPropsType = {
  name: string,
  size: number,
  color: string,
  style: any
}

export default {
  navButton (buttonProps: ButtonPropsType, callback: Function) {
    return (
      <TouchableOpacity onPress={callback}>
        <Icon {...defaultButtonProps} {...buttonProps} />
      </TouchableOpacity>
    )
  }
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: Colors.transparent,
    color: navButtonColor,
    justifyContent: 'center'
  }
})

const defaultButtonProps: ButtonPropsType = {
  name: 'heart',
  size: Metrics.icons.medium,
  color: Colors.snow,
  style: styles.button
}
