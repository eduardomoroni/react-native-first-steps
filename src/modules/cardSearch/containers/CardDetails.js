// @flow

import React, { Component } from 'react'
import { Text, View, ScrollView } from 'react-native'
import { styles } from '../../theme/cardSearch/CardDetailsStyle'
import { CardImage } from '../../shared/components/index'
import { cardType, rulingsType } from '../../../types/CardType'
import { rulingsTextAsArray } from '../../../services/realm/conversion/RealmRepresentation'
import I18n from 'react-native-i18n'

export default class CardDetails extends Component {
  renderText (value: string, key: number) {
    return (
      <Text key={key} >{`
        - ${value}`
      }</Text>
    )
  }

  renderRulings (rulings: rulingsType) {
    const rulesText = rulingsTextAsArray(rulings)

    return (
      <View style={styles.rowContainer}>
        <Text style={styles.rowText}>
          <Text style={styles.rowName}>{`${I18n.t('rulings')}: `}</Text>
          { rulesText.map(this.renderText) }
        </Text>
      </View>
    )
  }

  renderRow (rowName: string, rowText: string) {
    if (rowText === null) {
      return null
    }

    return (
      <View style={styles.rowContainer}>
        <Text style={styles.rowText}>
          <Text style={styles.rowName}>{`${I18n.t(rowName)}: `}</Text>
          {rowText}
        </Text>
      </View>
    )
  }

  render () {
    const { card } = this.props

    return (
      <ScrollView style={styles.container}>
        <View style={styles.card}>
          <CardImage card={card} />
        </View>
        {this.renderRow('name', card.name)}
        {this.renderRow('type', card.type)}
        {this.renderRow('P/T', `${card.power}/${card.toughness}`)}
        {this.renderRow('manaCost', `${card.manaCost} (${card.cmc})`)}
        {this.renderRow('rarity', card.rarity)}
        {this.renderRow('text', card.text)}
        {this.renderRow('flavor', card.flavor)}
        {this.renderRulings(card.rulings)}
        {this.renderRow('sets', 'Estou com pregui√ßa')}
        {this.renderRow('legality', 'Farei depois')}
        {this.renderRow('foreignNames', 'Um dia')}
        <View style={styles.card} />
      </ScrollView>
    )
  }
}

CardDetails.propTypes = {
  card: cardType
}
