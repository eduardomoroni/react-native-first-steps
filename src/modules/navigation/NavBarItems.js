import React from 'react'
import I18n from 'react-native-i18n'
import { Text, View } from 'react-native'
import { Actions as NavigationActions } from 'react-native-router-flux'
import { builtInBarStyle, customBarStyle } from '../theme/NavBarStyle'
import { NavButton } from './NavButton'

type ButtonPropsType = {
  name: string,
  size: number,
  style: any
}

// TODO: Enable Flow
// TODO: REFACTOR TYPES
export const navButton = (buttonProps: ButtonPropsType, callback: Function, key: any) => {
  return (
    <NavButton iconProps={{...buttonProps}} callback={callback} key={key} />
  )
}

export const openDrawer = (component: any) => {
  NavigationActions.refresh({
    key: 'drawer',
    content: component,
    open: true,
    side: 'left'
  })
}

export const navTitle = (title: string) => {
  return (
    <Text style={builtInBarStyle.titleStyle}>
      {I18n.t(title)}
    </Text>
  )
}

// TODO: This is an array of Object ButtonPropsType
export const navButtonBar = (buttons: any) => {
  return (
    <View style={[customBarStyle.rightButton, customBarStyle.buttons]} >
      { buttons.map(i => navButton(i.buttonProps, i.callback, i.buttonProps.name)) }
    </View>
  )
}

export const navBackButton = () => {
  return navButton({name: 'angle-left'}, NavigationActions.pop)
}
