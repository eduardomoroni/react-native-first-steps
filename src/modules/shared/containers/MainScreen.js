// @flow

import React, { Component } from 'react'
import {
  View,
  StyleSheet
} from 'react-native'
import { Colors } from '../modules/theme/themes'

export default class MainScreen extends Component {
  render () {
    return (
      <View style={styles.container} />
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: Colors.background
  }
})
