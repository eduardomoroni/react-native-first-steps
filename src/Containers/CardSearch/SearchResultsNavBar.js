import React, { Component } from 'react'
import { formValueSelector } from 'redux-form'
import { connect } from 'react-redux'
import { navBackButton, navTitle, navButtonBar, openDrawer } from '../../Navigation/NavBarItems'
import { switchDisplayMode as switchDisplayModeAction } from '../../Redux/Actions'
import Menu from '../CardSearch/ListCardFilterMenu'
import CustomNavBar from '../../Navigation/CustomNavBar'

class SearchResultsNavBar extends Component {
  render () {
    const { showCardsAs, switchDisplayMode } = this.props

    const settingsButton = {
      buttonProps: {name: 'sliders'},
      callback: () => openDrawer(<Menu />)
    }

    const displayModeButton = {
      buttonProps: {name: showCardsAs === 'image' ? 'th-list' : 'th-large'},
      callback: () => switchDisplayMode()
    }

    return (
      <CustomNavBar
        leftRender={navBackButton()}
        middleRender={navTitle('searching_cards')}
        rightRender={navButtonBar([displayModeButton, settingsButton])} />
    )
  }
}

const mapStateToProps = (state) => {
  const selector = formValueSelector('CardSearchFilter')
  return {
    showCardsAs: selector(state, 'showCardsAs')
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    switchDisplayMode: () => dispatch(switchDisplayModeAction())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultsNavBar)
