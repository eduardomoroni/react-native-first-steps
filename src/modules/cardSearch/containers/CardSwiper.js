import React, { Component } from 'react'
import { View } from 'react-native'
import PropTypes from 'prop-types'
import CardDetails from './CardDetails'
import { Actions as NavigationActions } from 'react-native-router-flux'
import { cardType } from '../../../types/CardType'
import { findCardIndex } from '../../../services/CardService'
import Swiper from 'react-native-swiper'
import { FloatingActionButton } from '../../shared/components/index'

const fabItems = [
  { name: 'md-share',
    buttonColor: '#9b59b6',
    title: 'Share',
    onPress: () => null
  }, {
    name: 'md-folder-open',
    buttonColor: '#3498db',
    title: 'Add to Deck',
    onPress: () => null
  }, {
    name: 'md-swap',
    buttonColor: '#1abc9c',
    title: 'Add to TradeList',
    onPress: () => null
  }
]

export class CardSwiper extends Component {
  renderCardDetails (card, key) {
    return (
      <CardDetails card={card} key={key} />
    )
  }

  updateScreenTitle (e, state) {
    const { index } = state
    const getCardName = (index) => this.children[index].props.card.name
    NavigationActions.refresh({title: `${getCardName(index)}`})
  }

  render () {
    const { cards, card } = this.props
    let { index } = this.props

    if (card !== null) {
      index = findCardIndex(cards, card)
    }

    return (
      <View>
        <Swiper
          ref={(component) => { this.swiper = component }}
          onMomentumScrollEnd={this.updateScreenTitle}
          style={{ flex: 1 }}
          loadMinimal
          loadMinimalSize={2}
          loop={false}
          showsPagination={false}
          index={index}
          >
          {cards.map(this.renderCardDetails)}
        </Swiper>
        <FloatingActionButton items={fabItems} />
      </View>
    )
  }
}

CardSwiper.propTypes = {
  cards: PropTypes.objectOf(cardType),
  card: cardType,
  index: PropTypes.number
}

CardSwiper.defaultProps = {
  index: 0,
  card: null
}

export default CardSwiper
