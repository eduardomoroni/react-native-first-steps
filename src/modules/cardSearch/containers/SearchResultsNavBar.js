// @flow

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { navBackButton, navTitle, navButtonBar, openDrawer } from '../../modules/navigation/NavBarItems'
import { switchDisplayMode as switchDisplayModeAction } from '../../redux/actions'
import Menu from './ListCardFilterMenu'
import CustomNavBar from '../../modules/navigation/CustomNavBar'
import type { Dispatch } from 'redux'

export class SearchResultsNavBar extends Component {
  render () {
    const { showCardsAs, switchDisplayMode } = this.props

    const settingsButton = {
      buttonProps: {name: 'sliders'},
      callback: () => openDrawer(<Menu />)
    }

    const displayModeButton = {
      buttonProps: {name: showCardsAs === 'image' ? 'th-list' : 'th-large'},
      callback: () => switchDisplayMode(showCardsAs === 'image' ? 'list' : 'image')
    }

    return (
      <CustomNavBar
        leftRender={navBackButton()}
        middleRender={navTitle('searching_cards')}
        rightRender={navButtonBar([displayModeButton, settingsButton])} />
    )
  }
}

const mapStateToProps = (state) => {
  const { showCardsAs } = state.cardSearch
  return {
    showCardsAs
  }
}

const mapDispatchToProps = (dispatch: Dispatch<*>) => {
  return {
    switchDisplayMode: (toggle) => dispatch(switchDisplayModeAction(toggle))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultsNavBar)
