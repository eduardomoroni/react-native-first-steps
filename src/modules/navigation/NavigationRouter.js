// @flow

import React from 'react'
import { Scene, Router } from 'react-native-router-flux'
import Settings from '../settings/containers/Settings'
import LoginScreen from '../login/containers/LoginScreen'
import MainScreen from '../shared/containers/MainScreen'
import Drawer, { menuDrawerProps } from './NavigationDrawer'
import CardSearchForm from '../cardSearch/containers/CardSearchForm'
import ListCards from '../cardSearch/containers/ListCards'
import CardDetails from '../cardSearch/containers/CardDetails'
import CardSwiper from '../cardSearch/containers/CardSwiper'
import WishList from '../cardTrade/containers/WishList'
import SearchResultsNavBar from '../cardSearch/containers/SearchResultsNavBar'
import I18n from 'react-native-i18n'
import { findCardsFromForm } from '../../services/CardService'
import { builtInBarStyle } from '../theme/NavBarStyle'

const testForm = {
  cardName: 'aetASDASDASD'
}

const testScreen = () => {
  const cards = findCardsFromForm(testForm)
  console.log(cards)
  return <WishList />
}

const NavigationRouter = () => {
  return (
    <Router>
      <Scene key='drawer' component={Drawer} {...menuDrawerProps} >
        <Scene key='drawerChildrenWrapper' {...builtInBarStyle} >
          <Scene key='cardSearchForm' component={CardSearchForm} title={I18n.t('card_search_form_title')} />
          <Scene key='cardDetails' component={CardDetails} />
          <Scene key='cardSwiper' component={CardSwiper} />
          <Scene key='wishList' component={WishList} />
          <Scene initial key='testScreen' component={testScreen} title='Test Screen' />
          <Scene key='mainScreen' component={MainScreen} title={I18n.t('welcome')} />
          <Scene key='settings' component={Settings} title={I18n.t('settings')} />
          <Scene key='listCards' component={ListCards} title={I18n.t('list_cards_title')} navBar={SearchResultsNavBar} />
          <Scene key='loginScreen' component={LoginScreen} title={I18n.t('login_for_title')} hideNavBar />
        </Scene>
      </Scene>
    </Router>
  )
}

export default NavigationRouter
