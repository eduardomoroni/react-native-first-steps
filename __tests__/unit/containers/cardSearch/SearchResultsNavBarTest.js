import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import { SearchResultsNavBar } from '../../../../src/modules/cardSearch/containers/SearchResultsNavBar'
import * as NavBarItems from '../../../../src/modules/navigation/NavBarItems'
NavBarItems.openDrawer = jest.fn()
jest.mock('react-native-vector-icons/FontAwesome', () => 'Icon')

const props = {
  showCardsAs: 'list',
  switchDisplayMode: jest.fn()
}

describe('<SearchResultsNavBar />', () => {
  it('Should Render SearchResultsNavBar', () => {
    const wrapper = shallow(<SearchResultsNavBar {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
