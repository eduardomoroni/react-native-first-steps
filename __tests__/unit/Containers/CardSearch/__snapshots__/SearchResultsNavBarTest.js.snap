// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultsNavBar /> Should Render SearchResultsNavBar 1`] = `
<CustomNavBar
  leftRender={
    <NavButton
      callback={[Function]}
      iconProps={
        Object {
          "name": "angle-left",
        }
      }
    />
  }
  middleRender={
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontFamily": "beleren",
          "fontSize": 18,
        }
      }
    >
      I18n.t(searching_cards)
    </Text>
  }
  rightRender={
    <View
      style={
        Array [
          Object {
            "justifyContent": "flex-end",
          },
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          },
        ]
      }
    >
      <NavButton
        callback={[Function]}
        iconProps={
          Object {
            "name": "th-large",
          }
        }
      />
      <NavButton
        callback={[Function]}
        iconProps={
          Object {
            "name": "sliders",
          }
        }
      />
    </View>
  }
/>
`;
